File: LeetCode/Medium/0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        if len(s) < 2 or s == s[::-1]:
            return s
        answer = ''
        for left in range(len(s) - 1):
            answer = max(answer,
                         self.expand(s, left, left + 1),
                         self.expand(s, left, left + 2),
                         key=len)
        return answer

    def expand(self, s: str, left: int, right: int) -> str:
        while left > -1 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        return s[left + 1: right]

File: LeetCode/Medium/0005-longest-palindromic-substring/0005-longest-palindromic-substring.kt
class Solution {

    var start = 0
    var end = 0

    fun longestPalindrome(s: String): String {
        for (i in 0 until s.lastIndex) {
            findPalindrome(s, i, i + 1)
            findPalindrome(s, i - 1, i + 1)
        }

        return s.substring(start..end)
    }

    fun findPalindrome(s: String, leftStart: Int, rightStart: Int) {
        var left = leftStart
        var right = rightStart

        while (left in s.indices && right in s.indices && s[left] == s[right]) {
            left--
            right++
        }

        left++
        right--

        if (end - start < right - left) {
            start = left
            end = right
        }
    }
}
